### ISSUE TITLE
New Ambientum CLI Planning

### ISSUE THREAD

**Issue by hernandev** [codecasts/ambientum#79]
### 1 - What.
Hi, this is a place where I want to share progress and get some feedback on the new command line interface for Ambientum.

I've quick started it's development and got some ideas, so feedback is always welcomed:

### 1 - General Requirements.

One idea is that Ambientum CLI would be able to init a project to use it. it means interactive docker-compose.yml generation.

Other idea is that it would replace project creation and other related binaries.

Ambientum since the beggining aims to replace local installed dev tools with Docker only versions.

So, considering the new CLI would need to:

- **Run PHP and Node commands on docker**
  Considering a `a-php` and `a-node` binaries that would be alternative to local installed but keeping compatible with local installed ones.
- **Be multi platform without custom config**
  It includes running on windows without much efforts, considering docker is operational on that machine.
- **Project initialization Helper**
  `amb init` would take care or generating the needed `docker-compose.yml` configuration file for a project, interactively asking about the components to include and customizations.
- **Project sensitive environment**
  Ambientum has it's own volumes to cache stand alone runs but being able to to tap into the project services is a must, most people using it open issued about running commands on the project. so `a-php composer install` would run using the project's links, like MySQL and Redis instead of isolated environment.

### 2 - Approach

Considering today we use a hard to maintain approach, shell scripts, I would move to another langue to develop this.

The options are:

- **PHP itself with Symfony Console**
  Hard to handle signal and terminal forwarding, so doing a `docker run` from within PHP code and keeping the code alive and the terminal working for the user is hard as hell.
  Also, hard do distribute in a multi-platform way without requiring previous PHP install.
- **Node.JS with PKG**
  Using Node would be easy to maintain, easy to forward commands, little to no tweaks on terminal handling but relying on docker tools, like `docker run` and abstracting some OS differences.
  The main downside on Node with PKG is that a final binary is somewhat 35MB, and UPX cannot reduce it's size. Since we're talking about 2 or 3 binaries, asking a user to download 75MB to start using this is not ideal. and makes distribution little harder.
- **Keep on .SH**
  That's a NO, too hard to keep shell scripts for each platform and test on it all.
- **Go(lang)**
  So far the best choice, cause it's possible to use the Docker source code itself to interact with Docker. No more guessing the binaries options and creating long run lines, since it's direct access.
  Although my experience with Go is not that great, I'm being able to deliver most of the requirements on Go.
  Also, final binaries are < 3MB after UPX compress, making distribution easier.
  That's my choice for now.

**Comment by mauri870** [comment]
Python is an alternative too, it interfaces directly with the Docker Engine API, but for CLI's and docker API Go is the best choice. If you need help I've worked with Go in the last two years so far and maybe can contribute to the project.

**Comment by hernandev** [comment]
@mauri870 Yeah help is always needed, the thing is this will probably produce two or 3 binaries, so using python relies on using PyInstaller which will bundle python alongside the actual code, producing at least 15MB binary each. not very friendly for me and the very same reason I don't go with Node.JS.

@mauri870 I will share the actual code progress on the CLI over the weekend so you can go over it,

I'm little experienced with Go, but made some good progress already, final binaries after [upx](https://github.com/upx/upx) are <3MB so that's really easy to distribute and self-upgrade

**Comment by pineli** [comment]
Excelent idea. My suggestion would be to do something like pg-cli <https://github.com/dbcli/pgcli>. 
I don't know what language it was made in, but it does have a very nice self-completion.

**Comment by hernandev** [comment]
This is 95% done, still not pushed, i'll update here once I finish
